src/grammar/block.rs

```diff
@@ -0,0 +1,415 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE.md
+// SPDX-FileCopyrightText: 2023 CYBERTEC PostgreSQL International GmbH
+// <office@cybertec.at>
+
+//! Implements parsing of SQL blocks from a token tree.
+
+use crate::grammar::{
+    expr, opt_expr, opt_function_invocation, opt_insert, opt_query, parse_datatype, parse_ident,
+};
+use crate::lexer::TokenKind;
+use crate::parser::Parser;
+use crate::syntax::SyntaxKind;
+use crate::ParseError;
+
+/// Looks ahead and parses a block if applicable
+pub(crate) fn opt_block(p: &mut Parser) -> bool {
+    if p.at(TokenKind::DeclareKw) || p.at(TokenKind::BeginKw) {
+        parse_block(p);
+        return true;
+    }
+    false
+}
+
+/// Parses a complete block.
+pub fn parse_block(p: &mut Parser) {
+    p.start(SyntaxKind::Block);
+    parse_declare_section(p);
+
+    p.expect(TokenKind::BeginKw);
```

Comment:
This will fail when the current token is the `declare` keyword accepted by the `opt_block` function.

Reply:
Never mind, I'm mistaken. Function `parse_declare_section` will parse the `declaration` section first.